eth_tx.go

```diff
@@ -0,0 +1,168 @@
+package ipldeth
+
+import (
+	"bytes"
+	"fmt"
+
+	cid "github.com/ipfs/go-cid"
+	node "github.com/ipfs/go-ipld-format"
+
+	types "github.com/ethereum/go-ethereum/core/types"
+	rlp "github.com/ethereum/go-ethereum/rlp"
+)
+
+// EthTx (eth-tx codec 0x93) represents an ethereum transaction
+type EthTx struct {
+	*types.Transaction
+
+	cid     *cid.Cid
+	rawdata []byte
+}
+
+// Static (compile time) check that EthTx satisfies the node.Node interface.
+var _ node.Node = (*EthTx)(nil)
+
+/*
+  INPUT
+*/
+
+// NewTx computes the cid and rlp-encodes a types.Transaction object
+// returning a proper EthTx node
+func NewTx(t *types.Transaction) *EthTx {
+	buf := new(bytes.Buffer)
+	if err := t.EncodeRLP(buf); err != nil {
+		panic(err)
+	}
+	rawdata := buf.Bytes()
+
+	return &EthTx{
+		Transaction: t,
+		cid:         rawdataToCid(MEthTx, rawdata),
+		rawdata:     rawdata,
+	}
+}
+
+/*
+ OUTPUT
+*/
+
+// DecodeEthTx takes a cid and its raw binary data
+// from IPFS and returns an EthTx object for further processing.
+func DecodeEthTx(c *cid.Cid, b []byte) (*EthTx, error) {
+	var t types.Transaction
+	err := rlp.DecodeBytes(b, &t)
+	if err != nil {
+		return nil, err
+	}
+
+	return &EthTx{
+		Transaction: &t,
+		cid:         c,
+		rawdata:     b,
+	}, nil
+}
+
+/*
+  Block INTERFACE
+*/
+
+// RawData returns the binary of the RLP encode of the transaction.
+func (t *EthTx) RawData() []byte {
+	return t.rawdata
+}
+
+// Cid returns the cid of the transaction.
+func (t *EthTx) Cid() *cid.Cid {
+	return t.cid
+}
+
+// String is a helper for output
+func (t *EthTx) String() string {
+	return fmt.Sprintf("<EthereumTx %s>", t.cid)
+}
+
+// Loggable returns in a map the type of IPLD Link.
+func (t *EthTx) Loggable() map[string]interface{} {
+	return map[string]interface{}{
+		"type": "eth-tx",
+	}
+}
+
+/*
+  Node INTERFACE
+*/
+
+// Resolve resolves a path through this node, stopping at any link boundary
+// and returning the object found as well as the remaining path to traverse
+func (t *EthTx) Resolve(p []string) (interface{}, []string, error) {
+	if len(p) == 0 {
+		return t, nil, nil
+	}
+
+	switch p[0] {
+	case "nonce":
+		return t.Nonce(), p[1:], nil
```

Comment:
As above, should return an error if `len(p) > 1`. We only use the second return value for links.